DelayedMessages/explainer.md

```diff
@@ -0,0 +1,757 @@
+      "processingEnd": 386.69999980926514,
+      "blockedDuration": 60,
+      "serializeDuration": 0,
+      "deserializeDuration": 0,
+      "scripts": [
+          {
+              "name": "script",
+              "entryType": "script",
+              "startTime": 213.30000019073486,
+              "duration": 121,
+              "invoker": "DedicatedWorkerGlobalScope.onmessage",
+              "invokerType": "event-listener",
+              "windowAttribution": "other",
+              "executionStart": 213.30000019073486,
+              "forcedStyleAndLayoutDuration": 0,
+              "pauseDuration": 0,
+              "sourceURL": "https://joone.github.io/web/explainers/delayed_messages/long_task/worker.js",
+              "sourceFunctionName": "runLongTaskOnWorker",
+              "sourceCharPosition": 168,
+              "sourceLine": 8,
+              "sourceColumn": 41
+          }
+      ],
+      "invoker": {
+          "name": "invoker",
+          "sourceURL": "https://joone.github.io/web/explainers/delayed_messages/long_task/main.js",
+          "sourceFunctionName": "sendMessage",
+          "sourceCharPosition": 510,
+          "sourceColumnNumber": 14,
+          "sourceLineNumber": 15,
+          "executionContext": {
+              "name": "",
+              "type": "window",
+              "id": 0
+          }
+      },
+      "receiver": {
+          "name": "receiver",
+          "sourceURL": "https://joone.github.io/web/explainers/delayed_messages/long_task/worker.js",
+          "sourceFunctionName": "runLongTaskOnWorker",
+          "sourceCharPosition": 168,
+          "sourceColumnNumber": 41,
+          "sourceLineNumber": 8,
+          "executionContext": {
+              "name": "",
+              "type": "dedicated-worker",
+              "id": 1
+          }
+      }
+  }
+]
+```
+
+# Alternatives Considered
+
+## DevTools Tracing
+
+Modern browser developer tools offer tracing capabilities that can show `postMessage` events and task execution timelines. While useful for manual inspection, these tools are not designed for systematic, large-scale metric collection, aggregation, or automated performance monitoring in production environments.
+
+The proposed API offers structured timing data suitable for production monitoring. It enables automated monitoring and performance attribution without requiring manual tooling, developer intervention, or complex trace analysis.
+
+## Manual Instrumentation / Polyfills
+
+Developers can attempt to wrap `postMessage` calls and `onmessage` handlers with their own timing logic (e.g., using `performance.now()`). However, this approach has several drawbacks:
+
+  * It's challenging to intercept all messages, especially those from third-party libraries.
+  * Accurately measuring internal browser operations like serialization, deserialization, and precise queue waiting time is not feasible from JavaScript.
+  * It adds boilerplate code and maintenance overhead.

```

Comment:
I think this is somewhat flipped. "Maintenance overhead" is the main reason to push back against adding new APIs to the web, as they become a permanent burden on what's already an extremely complex system.

If we could show that there are many websites out there doing manual instrumentation of `postMessage` and they would all benefit, that would strengthen the case for this proposal. (it might also help us shape the API as we'd be able to determine what commonalities the various clients all have/require)

Reply:
> I think this is somewhat flipped. "Maintenance overhead" is the main reason to push back against adding new APIs to the web, as they become a permanent burden on what's already an extremely complex system.
> 
 Okay, I’ll remove that part.

> If we could show that there are many websites out there doing manual instrumentation of `postMessage` and they would all benefit, that would strengthen the case for this proposal. (it might also help us shape the API as we'd be able to determine what commonalities the various clients all have/require)

Yes, we can look for examples of manual postMessage instrumentation both within Microsoft and across public websites.